core/trino-main/src/main/java/io/trino/execution/buffer/ExternalExchangeOutputBuffer.java

```diff
@@ -0,0 +1,249 @@
+            OutputBuffers outputBuffers,
+            ExchangeSink exchangeSink,
+            Supplier<LocalMemoryContext> systemMemoryContextSupplier)
+    {
+        this.state = requireNonNull(state, "state is null");
+        this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+        checkArgument(outputBuffers.getType() == EXTERNAL, "Expected an EXTERNAL output buffer");
+        this.exchangeSink = requireNonNull(exchangeSink, "exchangeSink is null");
+        this.systemMemoryContextSupplier = requireNonNull(systemMemoryContextSupplier, "systemMemoryContextSupplier is null");
+
+        state.compareAndSet(OPEN, NO_MORE_BUFFERS);
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        BufferState state = this.state.get();
+        return new OutputBufferInfo(
+                "EXTERNAL",
+                state,
+                false,
+                state.canAddPages(),
+                exchangeSink.getSystemMemoryUsage(),
+                totalPagesAdded.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+
+        // ignore buffers added after query finishes, which can happen when a query is canceled
+        // also ignore old versions, which is normal
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        // no more buffers can be added but verify this is valid state change
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void acknowledge(OutputBuffers.OutputBufferId bufferId, long token)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void abort(OutputBuffers.OutputBufferId bufferId)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ListenableFuture<Void> isFull()
+    {
+        return asVoid(toListenableFuture(exchangeSink.isBlocked()));
+    }
+
+    @Override
+    public void enqueue(List<SerializedPage> pages)
+    {
+        enqueue(0, pages);
```

Comment:
can you assert here that there is only one partition defined?

Reply:
Number of partitions that the sink is configured to accept is not directly available. In practice it will probably be encoded in the `ExchangeSinkInstanceHandle` and the `ExchangeSink` will be verifying it internally.